src/software/backend/simulation/simulator_robot_singleton.h

```diff
@@ -0,0 +1,257 @@
+#pragma once
+
+#include <cinttypes>
+#include <optional>
+
+#include "software/backend/simulation/simulator_robot.h"
+extern "C"
+{
+#include "firmware/main/app/world/chicker.h"
+#include "firmware/main/app/world/dribbler.h"
+#include "firmware/main/app/world/firmware_robot.h"
+#include "firmware/main/app/world/wheel.h"
+#include "firmware/main/shared/physics.h"
+}
+
+/**
+ * Because the FirmwareRobot_t struct is defined in the .c file (rather than the .h file),
+ * C++ considers it an incomplete type and is unable to use it with smart pointers
+ * because it doesn't know the size of the object. Therefore we need to create our own
+ * "Deleter" class we can provide to the smart pointers to handle that instead.
+ *
+ * See https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr for more info and
+ * examples
+ */
+struct FirmwareRobotDeleter
+{
+    void operator()(FirmwareRobot_t* firmware_robot) const
+    {
+        Wheel_t* front_left_wheel = app_firmware_robot_getFrontLeftWheel(firmware_robot);
+        app_wheel_destroy(front_left_wheel);
+
+        Wheel_t* back_left_wheel = app_firmware_robot_getBackLeftWheel(firmware_robot);
+        app_wheel_destroy(back_left_wheel);
+
+        Wheel_t* back_right_wheel = app_firmware_robot_getBackRightWheel(firmware_robot);
+        app_wheel_destroy(back_right_wheel);
+
+        Wheel_t* front_right_wheel =
+            app_firmware_robot_getFrontRightWheel(firmware_robot);
+        app_wheel_destroy(front_right_wheel);
+
+        Chicker_t* chicker = app_firmware_robot_getChicker(firmware_robot);
+        app_chicker_destroy(chicker);
+
+        Dribbler_t* dribbler = app_firmware_robot_getDribbler(firmware_robot);
+        app_dribbler_destroy(dribbler);
+
+        ControllerState_t* controller_state =
+            app_firmware_robot_getControllerState(firmware_robot);
+        delete controller_state;
+
+        app_firmware_robot_destroy(firmware_robot);
+    };
+};
+
+
+/**
+ * This class acts as a wrapper around a SimulatorRobot so that the SimulatorRobot
+ * can provide the interface of a FirmwareRobot.
+ *
+ * Because our firmware structs rely on C-style function pointers, we
+ * cannot use C++ constructs like std::function to provide to the structs.
+ * Therefore we use this class's static functions to provide the function
+ * pointers, since static functions work as C-style function pointers
+ * (because there is no instance associated with them).
+ *
+ * Whenever a caller needs to get a firmware struct or perform an operation, they
+ * need to set which robot they want to control so that all the static functions
+ * that have been provided to the firmware struct operate on the correct
+ * instantiated object. This is our workaround to maintain and simulate multiple
+ * "instances" of robot firmware at once.
+ */
```

Comment:
I don't really understand what's being said here. A few questions:
Who is the caller?
Why do we need to simulate multiple instances of firmware at once?
Is there any link where I can read more about this problem/solution?

Reply:
All good questions
> Why do we need to simulate multiple instances of firmware at once?

If we want to simulate an entire team of friendly robots, we need to simulate a "separate" instance of firmware for each, since firmware is written in a way that assumes it's running on a single robot.

> Who is the caller?
The caller would be the top-level `Simulator` object or whatever is in charge of managing the `SimulatorRobots` and simulating their corresponding primitives. This should be more clear in my next PR that is supposed to integrate everything

Here's an example from my current WIP code
```
void Simulator::stepSimulation(const Duration& time_step)
{
    SimulatorBallSingleton::setSimulatorBall(simulator_ball);
    for (auto iter = simulator_robots.begin(); iter != simulator_robots.end(); iter++)
    {
        auto simulator_robot = iter->first;
        auto primitive_manager = iter->second;
        SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);
        app_primitive_manager_runCurrentPrimitive(primitive_manager.get(), firmware_world.get());
    }

    physics_world.stepSimulation(time_step);
}
```
We set which simulator_robot we're "controlling" with the singleton, and then call the firmware functions. In this loop we are "swapping out" the simulator_robot being used each time we call the firmware functions, effectively letting us simulate multiple robots' firmware.

Hopefully that's a bit more clear, but please lmk if there's anything we can add to make it more understandable, since this is a pretty important piece.

As for reading, there isn't really a design pattern like this to my knowledge, we came up with this knowing a few things:
* C-style function pointers are different than c++ lambdas or `std::function` and they cannot be used together
* You cannot create C-style function pointers to member functions, only to static functions (because there's no instance associated with it) or normal functions (with no associated class)

https://isocpp.org/wiki/faq/pointers-to-members

Reply:
Another point I didn't mention explicitly.

1. We need to provide C-style function pointers to create a `FirmwareRobot_t` (which is needed to simulate firmware)
2. We cannot provide the member functions of a `SimulatorRobot` directly, because they are member functions and cannot be used as C-style function pointers
3. Therefore we need a "wrapper" that wraps a `SimulatorRobot` and exposes static functions, which can be used as C-style functions pointers. This is why the `SimulatorRobotSingleton` exists.

Happy to also explain in person on Saturday.